README.txt

this lib comes without any doku an will be used by some extensions of the author.

this extensions will not be supported but you can report bugs at http://bugs.naegler.net

ChangeLog
(add new changes on top of this file)
XX-XX-XX Frank Nägler  <typo3@naegler.net>
	* Bugfix: class.tx_fnlib_base.php->generateLanguageMarker | wrong encodings
	* NEW: class.tx_fnlib_twitter.php->enableNotification()
	* NEW: class.tx_fnlib_twitter.php->disableNotification()
	* NEW: class.tx_fnlib_twitpic.php

07-12-26 Frank Nägler  <typo3@naegler.net>
	* Initial release


lib/class.tx_fnlib_base.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2007 Frank Nägler <typo3@naegler.net>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

define(LOGLEVEL_INFO, 1);
define(LOGLEVEL_NOTICE, 2);
define(LOGLEVEL_WARNING, 4);
define(LOGLEVEL_ERROR, 8);
define(LOGLEVEL_FATAL, 16);
define(LOGLEVELS, array(
	0,
	LOGLEVEL_INFO,
	LOGLEVEL_NOTICE,
	LOGLEVEL_WARNING,
	LOGLEVEL_ERROR,
	LOGLEVEL_FATAL
));

/**
 * Class 'tx_fnlib_base' for the 'fn_lib' extension.
 *
 * @author	Frank Nägler <typo3@naegler.net>
 * @package	TYPO3
 * @subpackage	fn_lib
 */
class tx_fnlib_base {
	var $classes = array(
		'tx_fnlib_twitter'		=>	'lib/class.tx_fnlib_twitter.php',
		'tx_fnlib_twitpic'		=>	'lib/class.tx_fnlib_twitpic.php',
		'tx_fnlib_hookHandler'	=>	'lib/class.tx_fnlib_hookHandler.php'
	);

	/**
	 * contructur methode for this class
	 *
	 * @return tx_fnlib_base
	 */
	function tx_fnlib_base() {
		$this->init();
	}

	/**
	 * this methode make some inititial things like set global vars
	 * or include all needed classes.
	 *
	 */
	function init() {
		$this->EXT_PATH = t3lib_extMgm::extPath('fn_lib');
		foreach ($this->classes as $k => $v) {
			t3lib_div::requireOnce($this->EXT_PATH . $this->classes[$k]);
		}
	}
	
	/**
	 * set an logging level
	 *
	 * @param string $level
	 * @return void
	 */
	function setLogLevel($level=0) {
		if (in_array($level, LOGLEVELS)) {
			$this->logLevel = $level;
		} else {
			$this->logLevel = 0;
		}
	}
	
	/**
	 * send sysLog request if no loglevel is set or if a global loglevel is set
	 * than only if the given loglevel $level is lower as the global loglevel
	 *
	 * @param mixed $msg string or array which should be logged
	 * @param string $extKey
	 * @param string $level [fatal, error, warning, notive, info]
	 */
	function log($msg, $extKey, $level = '') {
		if (in_array($level, LOGLEVELS)) {
			$errLevel = $level;
		} else {
			$errLevel = 0;
		}
		
		if (is_array($msg)) {
			$msg = t3lib_div::arrayToLogString($msg);
		}
		
		if (isset($this->logLevel) && ($errlevel < $this->logLevel)) {
			return;
		}
		t3lib_div::sysLog($msg, $extKey, $errlevel);
	}
	
	/**
	 * get config settings from flexform
	 *
	 * @param object $pObj
	 * @return array $conf
	 */
	function getFlexformConf(&$pObj) {
		$piFlexForm = $pObj->cObj->data['pi_flexform'];
		foreach ( $piFlexForm['data'] as $sheet => $data ) {
			foreach ( $data as $lang => $value ) {
				foreach ( $value as $key => $val ) {
					$conf[$key] = $pObj->pi_getFFvalue($piFlexForm, $key, $sheet);
				}
			}
		}
		return $conf;
	}
	
	/**
	 * this methode get back an instance of the
	 * twitter client
	 *
	 * @return object	tx_fnlib_twitter instance
	 */
	function getTwitterClient($user='', $pass='') {
		return $this->twitterClient = new tx_fnlib_twitter($user, $pass);
	}

	/**
	 * this methode get back an instance of the
	 * twitter client
	 *
	 * @return object	tx_fnlib_twitter instance
	 */
	function getTwitpicClient($user, $pass) {
		return $this->twitpicClient = new tx_fnlib_twitpic($user, $pass);
	}

	/**
	 * this methode get back an instance of the
	 * hookHandler class
	 *
	 * @return object	tx_fnlib_hookHandler instance
	 */
	function getHookHandler($extKey) {
		// return $this->hookHandlert = t3lib_div::makeInstance($this->classes['tx_fnlib_hookHandler']);
		return $this->hookHandlert = new tx_fnlib_hookHandler($extKey);
	}
	
	/**
	 * Gets all "lang_ and label_" Marker for substition with substituteMarkerArray
	 * Take function from tx_commerce
	 * @return      void
	 * @since 10.02.06 Changed to XML
	 * @coauthor Frank Kroeber <fk@marketing-factory.de>
	 * @coauthor Frank Nägler <typo3@naegler.net>
	 */
	function generateLanguageMarker(&$pObj) {
		$pObj->languageMarker = array();
		if (
			(is_array($pObj->LOCAL_LANG[$GLOBALS['TSFE']->tmpl->setup['config.']['language']]))
			&&
			(is_array($pObj->LOCAL_LANG['default']))
		) {
			$markerArr = array_merge($pObj->LOCAL_LANG['default'],
			$pObj->LOCAL_LANG[$GLOBALS['TSFE']->tmpl->setup['config.']['language']]);
		} elseif (is_array($pObj->LOCAL_LANG['default'])) {
			$markerArr=$pObj->LOCAL_LANG['default'];
		} else {
			$markerArr=$pObj->LOCAL_LANG[$GLOBALS['TSFE']->tmpl->setup['config.']['language']];
		}
		
		while(list($k,$v)=each($markerArr)) {
			if (stristr($k,'lang_') OR stristr($k,'label_')) {
				$pObj->languageMarker['###'.strtoupper($k).'###'] = $pObj->pi_getLL($k);
			}
		}
	}
	
	/**
	 * generates a markerArray by given table and uid or where clause
	 *
	 * @param string 	$table
	 * @param int		$uid
	 * @param string	$where
	 * @return array
	 */
	function generateTableMarker($table, $uid = 0, $where = '') {
		$tableMarker = array();
		if ($uid == 0) {
			$res = $GLOBALS['TYPO3_DB']->exec_SELECTquery(
				'*',
				$table,
				$where
			);
		} else {
			$res = $GLOBALS['TYPO3_DB']->exec_SELECTquery(
				'*',
				$table,
				'uid = ' . $uid . $where
			);
		}
		if ($GLOBALS['TYPO3_DB']->sql_num_rows($res) == 1) {
			$data = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($res);
			foreach ($data as $k => $v) {
				$tableMarker['###' . strtoupper($table) . '_' . strtoupper($k) . '###'] = $data[$k];
			}
		} else if ($GLOBALS['TYPO3_DB']->sql_num_rows($res) > 1) {
			while ($data = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($res)) {
				$tempMarker = null;
				foreach ($data as $k => $v) {
					$tempMarker['###' . strtoupper($table) . '_' . strtoupper($k) . '###'] = $data[$k];
				}
				$tableMarker[] = $tempMarker;
			}
		}
		return $tableMarker;
	}
	
	/**
	 * this methode parse the given markerArray into the subpart
	 * of the given template and return the parsed template code.
	 *
	 * @param array		$conf
	 * @param string	$subpart
	 * @param array		$markerArray
	 * @param object	$pObj
	 * @return string
	 */
	function parseTemplate(&$conf, $subpart, &$markerArray, &$pObj) {
		$template = $pObj->cObj->fileResource($conf['template']);
		$templateCode = $pObj->cObj->getSubpart($template, '###TEMPLATE_'.strtoupper($subpart).'###');
		return $pObj->cObj->substituteMarkerArray($templateCode, $markerArray);
	}
	
	/**
	 * get TypoScript config array by given key
	 *
	 * @param string $key
	 * @return array
	 */
	function getTypoScript4PlugIn($key) {
		return $GLOBALS['TSFE']->tmpl->setup['plugin.'][$key.'.'];
	}

}

?>


lib/class.tx_fnlib_twitpic.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2008 Frank Nägler <typo3@naegler.net>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * Class 'tx_fnlib_twitpic' for the 'fn_lib' extension.
 *
 * @author	Frank Nägler <typo3@naegler.net>
 * @package	TYPO3
 * @subpackage	fn_lib
 */
class tx_fnlib_twitpic {
	var $host = 'http://twitpic.com';
	var $actions = array(
		'uploadAndPost' => '/api/uploadAndPost',
		'upload'		=> '/api/upload',
	);

	var $status = array(
		'1001' => 'Invalid twitter username or password',
		'1002' => 'Image not found',
		'1003' => 'Invalid image type',
		'1004' => 'Image larger than 4MB ',
	);

	function tx_fnlib_twitpic($username, $password) {
		$this->username = $username;
		$this->password = $password;
		$this->init();
	}

	function init() {
		$this->transporter = curl_init();
		curl_setopt($this->transporter, CURLOPT_HEADER, true);
		curl_setopt($this->transporter, CURLOPT_USERPWD, $this->username . ":" . $this->password);
		curl_setopt($this->transporter, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
		curl_setopt($this->transporter, CURLOPT_RETURNTRANSFER, true);
		curl_setopt($this->transporter, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
		curl_setopt($this->transporter, CURLOPT_HEADERFUNCTION, array($this, '_setHeader'));
	}

	function _setHeader($ch, $header) {
		$this->headers[] = $header;
        return strlen($header);
	}

	function makeRequest($format) {
		$this->headers = array();
		$status = curl_exec($this->transporter);
		return ($status) ? ($status) : ($this->checkFailure());
	}

	function checkFailure() {
		print_r($this->headers);
	}

	function uploadAndPost($media, $msg = '') {
		curl_setopt($this->transporter, CURLOPT_URL, $this->host . $this->actions['update']);
		$msg = str_replace('  ', ' ', $msg);
		$params = array(
			'media' => $media
		);
		curl_setopt($this->transporter, CURLOPT_POSTFIELDS, $params);
		return $this->makeRequest();
	}

	function upload($media) {
		return $this->uploadAndPost($media);
	}
}

?>


lib/class.tx_fnlib_twitter.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2007 Frank Nägler <typo3@naegler.net>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * Class 'tx_fnlib_twitter' for the 'fn_lib' extension.
 *
 * @author	Frank Nägler <typo3@naegler.net>
 * @package	TYPO3
 * @subpackage	fn_lib
 */
class tx_fnlib_twitter {
	var $host = 'http://twitter.com';
	var $actions = array(
		'update' => '/statuses/update',
		'timeline' => array(
			'public' => '/statuses/public_timeline',
			'friends' => '/statuses/friends_timeline',
			'user' => '/statuses/user_timeline'
		),
		'getStatus' => '/statuses/show/id',
		'replies' => '/statuses/replies',
		'destroy' => '/statuses/destroy',
		'friends' => '/statuses/friends',
		'followers' => '/statuses/followers',
		'featured' => '/statuses/featured',
		'userinfo' => '/users/show',
		'direct_messages' => '/direct_messages',
		'direct_messages_sent' => '/direct_messages/sent',
		'sent_direct_message' => '/direct_messages/new',
		'destroy_direct_message' => '/direct_messages/destroy',
		'create_friend' => '/friendships/create',
		'destroy_friend' => '/friendships/destroy',
		'verify_credentials' => '/account/verify_credentials',
		'end_session' => '/account/end_session',
		'archive' => '/account/archive',
		'favorites' => '/favourings',
		'create_favorites' => '/favourings/create',
		'destroy_favorites' => '/favourings/destroy',
		'notifications' => array(
			'follow' => '/notifications/follow',
			'leave' => '/notifications/leave',
		),
	);

	var $status = array(
		'200' => 'OK: everything went awesome.',
		'302' => 'Not Modified: there was no new data to return.',
		'400' => 'Bad Request:',
		'401' => 'Not Authorized:',
		'403' => 'Forbidden:',
		'404' => 'Not Found: ',
		'500' => 'Internal Server Error:',
		'502' => 'Bad Gateway: returned if Twitter is down or being upgraded.',
		'503' => 'Service Unavailable: the Twitter servers are up, but are overloaded with requests.  Try again later.'
	);

	function tx_fnlib_twitter($username, $password) {
		$this->username = $username;
		$this->password = $password;
		$this->init();
	}

	function init() {
		$this->transporter = curl_init();
		curl_setopt($this->transporter, CURLOPT_HEADER, true);
		curl_setopt($this->transporter, CURLOPT_USERPWD, $this->username . ":" . $this->password);
		curl_setopt($this->transporter, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
		curl_setopt($this->transporter, CURLOPT_RETURNTRANSFER, true);
		curl_setopt($this->transporter, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
		curl_setopt($this->transporter, CURLOPT_HEADERFUNCTION, array($this, '_setHeader'));
	}

	function _setHeader($ch, $header) {
		$this->headers[] = $header;
        return strlen($header);
	}

	function makeRequest($format) {
		$this->headers = array();
		$status = curl_exec($this->transporter);
		//debug($status);
		return ($status) ? ($status) : ($this->checkFailure($format));
	}

	function checkFailure($format) {
		print_r($this->headers);
	}

	/**
	 * get tiny url from tinyurl.com API
	 *
	 * @param string $url
	 * @return string
	 */
	function getTinyUrl($url) {
		$url = 'http://tinyurl.com/api-create.php?url=' . $url;
		$cu = curl_init();
		curl_setopt($cu, CURLOPT_RETURNTRANSFER, true);
		curl_setopt($cu, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
		curl_setopt($cu, CURLOPT_URL, $url);
		return curl_exec($cu);
	}
	
	// Status Methods
	function update($msg, $format = 'xml') {
		curl_setopt($this->transporter, CURLOPT_URL, $this->host . $this->actions['update'] . ".$format");
		$msg = str_replace('  ', ' ', $msg);
		$params = array(
			'status' => $msg
		);
		curl_setopt($this->transporter, CURLOPT_POSTFIELDS, $params);
		return $this->makeRequest($format);
	}

	function destroy($id, $format = 'xml') {
		$action = $this->host . $this->actions['destroy'] . '/' . intval($id) . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest($format);
	}

	function getTimeline($which = 'public', $format = 'xml', $params = null) {
		$action = $this->host . $this->actions['timeline'][$which] . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		if ($params != null)
			curl_setopt($this->transporter, CURLOPT_POSTFIELDS, $params);
		return $this->makeRequest($format);
	}

	function getStatus($id, $format = 'xml') {
		$action = $this->host . $this->actions['getStatus'] . '/' . intval($id) . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest($format);
	}

	function getReplies($format = 'xml', $params = null) {
		$action = $this->host . $this->actions['replies'] . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		if ($params != null)
			curl_setopt($this->transporter, CURLOPT_POSTFIELDS, $params);
		return $this->makeRequest($format);
	}

	// User Methods
	function getFriends($format = 'xml', $params = null) {
		$action = $this->host . $this->actions['friends'] . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		if ($params != null)
			curl_setopt($this->transporter, CURLOPT_POSTFIELDS, $params);
		return $this->makeRequest($format);
	}

	function getFollowers($format = 'xml', $lite = false) {
		$action = $this->host . $this->actions['followers'] . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		if ($lite)
			curl_setopt($this->transporter, CURLOPT_POSTFIELDS, array('lite' => 'true'));
		return $this->makeRequest($format);
	}

	function getFeatured($format = 'xml') {
		$action = $this->host . $this->actions['featured'] . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest($format);
	}

	function getUserinfoById($id, $format = 'xml') {
		$action = $this->host . $this->actions['userinfo'] . '/' . $id . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest($format);
	}

	function getUserinfoByEmail($email, $format = 'xml') {
		$action = $this->host . $this->actions['userinfo'] . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		curl_setopt($this->transporter, CURLOPT_POSTFIELDS, array('email' => $email));
		return $this->makeRequest($format);
	}

	// Direct Message Methods
	function getDirectMessages($format = 'xml', $params = null) {
		$action = $this->host . $this->actions['direct_messages'] . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		if ($params != null)
			curl_setopt($this->transporter, CURLOPT_POSTFIELDS, $params);
		return $this->makeRequest($format);
	}

	function getSentDirectMessages($format = 'xml', $params = null) {
		$action = $this->host . $this->actions['direct_messages_sent'] . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		if ($params != null)
			curl_setopt($this->transporter, CURLOPT_POSTFIELDS, $params);
		return $this->makeRequest($format);
	}

	function sentDirectMessage($user, $msg, $format = 'xml') {
		$action = $this->host . $this->actions['sent_direct_message'] . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		$params = array(
			'user' => $user,
			'text' => $msg
		);
		curl_setopt($this->transporter, CURLOPT_POSTFIELDS, $params);
		return $this->makeRequest($format);
	}

	function destroyDirectMessage($id, $format = 'xml') {
		$action = $this->host . $this->actions['destroy_direct_message'] . '/' . $id . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest($format);
	}
	
	// Friendship Methods
	function createFriend($id, $format = 'xml') {
		$action = $this->host . $this->actions['create_friend'] . '/' . $id . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest($format);
	}

	function destroyFriend($id, $format = 'xml') {
		$action = $this->host . $this->actions['destroy_friend'] . '/' . $id . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest($format);
	}
	
	// Account Methods
	function verify_credentials($format = null) {
		$action = $this->host . $this->actions['verify_credentials'];
		$action .= ($format != null) ? (".$format") : ('');
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest($format);
	}

	function end_session() {
		$action = $this->host . $this->actions['end_session'];
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest(null);
	}

	function getArchive($format = 'xml', $page = null) {
		$action = $this->host . $this->actions['archive'] . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		if ($page != null) {
			$params = array(
				'page' => intval($page)
			);
			curl_setopt($this->transporter, CURLOPT_POSTFIELDS, $params);
		}		
		return $this->makeRequest($format);
	}
	
	// Favorite Methods
	function getFavorites($format = 'xml', $params = null) {
		$action = $this->host . $this->actions['favorites'] . ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		if ($params != null)
			curl_setopt($this->transporter, CURLOPT_POSTFIELDS, $params);
		return $this->makeRequest($format);
	}

	function createFavorite($id, $format = 'xml') {
		$action = $this->host . $this->actions['create_favorites'] . '/' . intval($id). ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest($format);
	}

	function destroyFavorite($id, $format = 'xml') {
		$action = $this->host . $this->actions['destroy_favorites'] . '/' . intval($id). ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest($format);
	}

	// Notification Methods
	function enableNotification($id, $format = 'xml') {
		$action = $this->host . $this->actions['notications']['follow'] . '/' . $id. ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest($format);
	}

	function disableNotification($id, $format = 'xml') {
		$action = $this->host . $this->actions['notications']['leave'] . '/' . $id. ".$format";
		curl_setopt($this->transporter, CURLOPT_URL, $action);
		return $this->makeRequest($format);
	}

}

?>


lib/class.tx_fnlib_hookHandler.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2007 Frank Nägler <typo3@naegler.net>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/


/**
 * Class 'tx_fnlib_hookHandler' for the 'fn_lib' extension.
 *
 * @author	Frank Nägler <typo3@naegler.net>
 * @package	TYPO3
 * @subpackage	fn_lib
 */
class tx_fnlib_hookHandler {
	
	/**
	 * contructur methode for this class
	 *
	 * @return tx_fnlib_hookHandler
	 */
	function tx_fnlib_hookHandler($extKey) {
		$this->extKey = $extKey;
		$this->init();
	}
	
	/**
	 * this methode make some inititial things like set global vars
	 * or include all needed classes.
	 *
	 */
	function init() {
		$this->hookObjectsArr = array();
		// init Hooks for given Extension
		if (is_array ($GLOBALS['TYPO3_CONF_VARS']['EXTCONF'][$this->extKey]['hookHandler'])) {
			foreach ($GLOBALS['TYPO3_CONF_VARS']['EXTCONF'][$this->extKey]['hookHandler'] as $classRef) {
				$this->hookObjectsArr[] = &t3lib_div::getUserObj($classRef);
			}
		}
	}
	
	/**
	 * this method calls a method of a each registerd hookObj
	 *
	 * @param string $method
	 * @param mixed	some other params which will be transfered to the hook methods
	 * @return defined by called hook object method
	 */
	function call($method) {
		$params = func_get_args();
		array_shift($params);
		if(is_array($this->hookObjectsArr)){
			foreach($this->hookObjectsArr as $hookObj) {
				if (method_exists($hookObj, $method)) {
					return call_user_func_array(array(&$hookObj, $method), $params);
				}
			}
		}
	}
}
?>

ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext: "fn_lib"
#
# Auto generated 11-07-2007 22:48
#
# Manual updates:
# Only the data in the array - anything else is removed by next write.
# "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Frank Naegler libs',
	'description' => 'some libs',
	'category' => 'plugin',
	'author' => 'Frank Nägler',
	'author_email' => 'typo3@naegler.net',
	'shy' => '',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'alpha',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => '',
	'version' => '0.1.0',
	'constraints' => array(
		'depends' => array(
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:13:{s:9:"ChangeLog";s:4:"ca07";s:10:"README.txt";s:4:"9d9a";s:36:"class.tx_timtab_embeddedvideo_fe.php";s:4:"74e4";s:12:"ext_icon.gif";s:4:"8d87";s:17:"ext_localconf.php";s:4:"808e";s:14:"ext_tables.php";s:4:"ca76";s:13:"locallang.xml";s:4:"c6ac";s:16:"locallang_db.xml";s:4:"ce65";s:14:"doc/manual.sxw";s:4:"0941";s:19:"doc/wizard_form.dat";s:4:"06f0";s:20:"doc/wizard_form.html";s:4:"c966";s:44:"static/tx_timtab_embeddedvideo/constants.txt";s:4:"80ae";s:40:"static/tx_timtab_embeddedvideo/setup.txt";s:4:"d223";}',
	'suggests' => array(
	),
);

?>